nemo/collections/nlp/data/machine_translation/preproc_mt_data.py

```diff
@@ -250,107 +272,67 @@ def preprocess_parallel_dataset(
 
-                if num_files_in_tar != num_batches_per_tarfile:
-                    os.remove(
-                        os.path.join(
-                            out_dir, '%s-batches.tokens.%d.%d.tar' % (pkl_file_prefix, tokens_in_batch, tar_file_ctr)
-                        )
+                # get number of lines so that we can create a partition of the lines of the text file
+                num_src_lines, num_tgt_lines = Parallel(n_jobs=2)(
+                    delayed(MTDataPreproc._get_num_lines)(filename) for filename in filenames
+                )
+                logging.info(f'Found {num_src_lines} source lines and  {num_tgt_lines} target lines.')
+                assert num_src_lines == num_tgt_lines, 'Number of source lines should equal number of target lines.'
+
+                # create a partition of lines that we can parallelize over
+                lines_partition = MTDataPreproc._get_lines_partition(num_src_lines, lines_per_dataset_fragment)
+
+                # create tarfiles for each fragment in parallel
+                # total_batches = 0
+                total_batches_list = Parallel(n_jobs=n_jobs)(
+                    delayed(MTDataPreproc._process_fragment)(
+                        src_filename=src_fname,
+                        tgt_filename=tgt_fname,
+                        lines_indices=lines_indices,
+                        out_dir=out_dir,
+                        num_batches_per_tarfile=num_batches_per_tarfile,
+                        clean=clean,
+                        max_seq_length=max_seq_length,
+                        min_seq_length=min_seq_length,
+                        tokens_in_batch=tokens_in_batch,
+                        encoder_tokenizer_name=encoder_tokenizer_name,
+                        encoder_tokenizer_model=encoder_tokenizer_model,
+                        encoder_bpe_dropout=encoder_bpe_dropout,
+                        encoder_model_name=encoder_model_name,
+                        decoder_tokenizer_name=decoder_tokenizer_name,
+                        decoder_tokenizer_model=decoder_tokenizer_model,
+                        decoder_bpe_dropout=decoder_bpe_dropout,
+                        decoder_model_name=decoder_model_name,
+                        pkl_file_prefix=fragment_index,
                     )
-                    global_batch_ctr -= num_files_in_tar
-                    logging.info('Dropping %d batches because of overflow' % (num_files_in_tar))
+                    for fragment_index, lines_indices in enumerate(lines_partition)
+                )
+                total_batches = sum(total_batches_list)
 
-                json.dump({'num_batches': global_batch_ctr}, open(metadata_path, 'w'))
+                # dump metadata to json
+                metadata = {}
+                metadata['num_batches'] = total_batches
+                tar_file_paths = glob.glob(f'{out_dir}/*.tar')
+                metadata['tar_files'] = tar_file_paths
```

Comment:
Adding this path to the metadata may make moving an already created tarred dataset around tricky. Since you disallow `cfg.tar_files` and metadata['tar_files']. How do I override what's in the metadata to use what I provide via hydra?

Reply:
You can still specify your tarred dataset via hydra. But now you can also specify it via a manifest file.